api/src/main/java/io/grpc/MetricRecorder.java

```diff
@@ -33,7 +35,16 @@ public interface MetricRecorder {
    * @param optionalLabelValues A list of additional, optional label values for the metric.
    */
   default void addDoubleCounter(DoubleCounterMetricInstrument metricInstrument, double value,
-      List<String> requiredLabelValues, List<String> optionalLabelValues) {}
+      List<String> requiredLabelValues, List<String> optionalLabelValues) {
+    checkArgument(requiredLabelValues != null
+            && requiredLabelValues.size() == metricInstrument.getRequiredLabelKeys().size(),
+        "Incorrect number of required labels provided. Expected: "
+            + metricInstrument.getRequiredLabelKeys().size());
```

Comment:
This does string concatenation even when there isn't a failure. Yes, I see `MetricRecorderImpl` had the same problem. We will need to use `%s`-powered templates to format the message.

Since this is already broken, that can be a follow-up.